sympy/algebras/euler.py

```diff
@@ -0,0 +1,676 @@
+from mpmath.libmp.libmpf import prec_to_dps
+
+
+def _check_sequence(seq):
+    """validate seq and return info"""
+    if isinstance(seq, Symbol):
+        seq = str(seq)
+    if type(seq) != str:
+        raise ValueError('Expected seq to be a string.')
+    if len(seq) != 3:
+        raise ValueError("Expected 3 axes, got `{}`.".format(seq))
+
+    intrinsic = seq.isupper()
+    extrinsic = seq.islower()
+    if not (intrinsic or extrinsic):
+        raise ValueError("seq must either be fully uppercase (for extrinsic "
+                         "rotations), or fully lowercase, for intrinsic "
+                         "rotations).")
+
+    i, j, k = seq.lower()
+    if i == j or j == k:
+        raise ValueError("Consecutive axes must be different")
+
+    bad = set(seq) - set('xyzXYZ')
+    if bad:
+        raise ValueError("Expected axes from `seq` to be from "
+                         "['x', 'y', 'z'] or ['X', 'Y', 'Z'], "
+                         "got {}".format(''.join(bad)))
+
+    # get elementary basis vectors
+    ei = [1 if n == i else 0 for n in 'xyz']
+    ej = [1 if n == j else 0 for n in 'xyz']
+    ek = [1 if n == k else 0 for n in 'xyz']
+
+    # get indices
+    i = 'xyz'.index(i) + 1
+    j = 'xyz'.index(j) + 1
+    k = 'xyz'.index(k) + 1
+
+    info = {
+            'seq': seq,
+            'extrinsic': extrinsic,
+            'intrinsic': intrinsic,
+            'symmetric': i == k,
+            'i': i,
+            'j': j,
+            'k': k,
+            'ei': ei,
+            'ej': ej,
+            'ek': ek
+            }
+
+    return info
+
+
+def _wrap(angle):
+    if angle.is_number:
+        return (angle + pi) % (2 * pi) - pi
+    else:
+        return angle
+
+
+class Euler(Expr):
+    """Provides Euler angles operations.
+
+    Euler objects can be instantiated as Euler(alpha, beta, gamma, seq).
+
+    Parameters
+    ==========
+
+    seq : string of length 3
+        Represents the sequence of rotations.
+        For intrinsic rotations, seq must be all lowercase and its elements
+        must be from the set `{'x', 'y', 'z'}`
+        For extrinsic rotations, seq must be all uppercase and its elements
+        must be from the set `{'X', 'Y', 'Z'}`
+
+    References
+    ==========
+
+    .. [1] https://en.wikipedia.org/wiki/Euler_angles
+
+    """
+    _op_priority = 11.0
+
+    is_commutative = False
+
+    def __new__(cls, alpha=0, beta=0, gamma=0, seq='ZYX'):
+        info = _check_sequence(seq)
+        alpha, beta, gamma, seq= map(sympify, (alpha, beta, gamma, seq))
+        alpha, beta, gamma, seq = map(_wrap, (alpha, beta, gamma, seq))
+
+        if any(i.is_commutative is False for i in [alpha, beta, gamma]):
+            raise ValueError("arguments have to be commutative")
+        else:
+            obj = Expr.__new__(cls, alpha, beta, gamma, seq)
+            obj._alpha = alpha
+            obj._beta = beta
+            obj._gamma = gamma
+            obj._info = info
+            obj._q = None
```

Comment:
These attributes should not be set. Instead everything should be derived from `self.args`.

Reply:
Alright, did just that, and removed the `info` dict I had defined.

I left the `_q` attribute to lazily calculate the quaternion since it might give some speed advantages.

Moreover, `Euler.seq` gives `str(Euler.args[3])` instead of `Euler.args[3]`, this is because of the workaround I had to do to put `seq` as a `sympify`ed `Symbol` in `args`.